### PR TITLE
Added recursive fetch of child domains for listUsageRecords API call

### PR DESCRIPTION
**Pull request by Spaceman1984**
### Description

When calling the listUasageRecords API records per domain are fetched recursively. This is not the case if you specify a domain id.

This PR adds a new parameter to enable fetching records recursively (isRecursive) when passing the domain id.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### How Has This Been Tested?

This has been tested by calling the API with different parameters. With/without a domain id, with/without isRecursive and with/without an account id.
I have tested by calling the API logged in as a root admin, domain admin, and a user.

### CODE REVIEW

**Comment by Spaceman1984** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 1]
```diff
@@ -248,7 +248,17 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
             sc.addAnd("domainId", SearchCriteria.Op.IN, domainIds.toArray());
         }
 
-        if (domainId != null) {
+        if (cmd.isRecursive() && domainId != null){
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:251]: Look at for reference https://github.com/apache/cloudstack/blob/master/server/src/main/java/com/cloud/api/query/QueryManagerImpl.java#L872

Expected test if an account is allowed or disallowed to get usage records of other domains (with/without recursive flag), and allowed to recursively get all usage records for a permitted account/domain.

Comment (Spaceman1984) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:276]: While testing I found that users cannot call this API, but domain admins can. I have added a check, if an account id is specified, if it belongs to any of the domains for the calling domain admin.

**Comment by Spaceman1984** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 2]
```diff
@@ -216,6 +216,31 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
             s_logger.debug("Account details not available. Using userContext accountId: " + accountId);
         }
 
+        // Check if a domain admin is allowed to access the requested account info.
+        if (_accountService.isDomainAdmin(caller.getId()) && accountId != null){
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:220]: @Spaceman1984 What is the caller account is root admin?

Comment (rhtyd) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:220]: or normal user account?

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 3]
```diff
@@ -216,6 +216,31 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
+            boolean matchFound = false;
+
+            if (callerDomainId == accountDomainId) {
+                matchFound = true;
+            } else {
+                // Check if the account is in a child domain of this domain admin.
+                List<DomainVO> childDomains = _domainDao.findAllChildren(_domainDao.findById(caller.getDomainId()).getPath(), caller.getDomainId());
+
+                for (DomainVO domainVO: childDomains) {
+                    if (accountDomainId == domainVO.getId()) {
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:232]: Why not use the existing `_accountMgr.buildACLSearchParameters`?
And should the be done for all caller types, as: (this is just suggestion, I've not investigated this)
```
Root admin - allow recursive usage records for all domains
Domain admin - only allow recursive usage records for its own domain
User - don't allow recursive domain ID (it should be only allowed to see their own usage records)
```

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java (hunk 2 above)
Comment (Spaceman1984) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:220]: A normal user doesn't have access to this API, this check is only applicable to domain admin.

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java (hunk 3 above)
Comment (Spaceman1984) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:232]: The _accountMgr.buildACLSearchParameter method needs a seaarchBuilder and not searchCriteria. I believe you intended it for fetching usage records, but I'm I am simply doing authorization here. 

If a Root admin does the call with a domain id, that is handled on line 276. If a domain admin does the call, authorization will happen first. Users cannot call this API.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java [hunk 4]
```diff
@@ -85,6 +85,10 @@
     @Parameter(name = ApiConstants.OLD_FORMAT, type = CommandType.BOOLEAN, description = "Flag to enable description rendered in old format which uses internal database IDs instead of UUIDs. False by default.")
     private Boolean oldFormat;
 
+    @Parameter(name = ApiConstants.IS_RECURSIVE, type = CommandType.BOOLEAN,
+            description = "Specify if usage records should be fetched recursively per domain.")
```
Comment (shwstppr) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java:89]: @Spaceman1984 you may add the `since` key here

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java (hunk 3 above)
Comment (shwstppr) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:232]: @Spaceman1984 can we use checkAccess method to simplify checking accountId domain access?
`void checkAccess(Account account, Domain domain) throws PermissionDeniedException;`

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 5]
```diff
@@ -248,7 +273,17 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
+            SearchCriteria<DomainVO> sdc = _domainDao.createSearchCriteria();
+            sdc.addOr("path", SearchCriteria.Op.LIKE, _domainDao.findById(domainId).getPath() + "%");
+            List<DomainVO> domains = _domainDao.search(sdc, null);
+            List<Long> domainIds = new ArrayList<Long>();
+            for (DomainVO domain : domains)
+                domainIds.add(domain.getId());
+            sc.addAnd("domainId", SearchCriteria.Op.IN, domainIds.toArray());
+        }
+
+        if (!cmd.isRecursive() && domainId != null) {
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:286]: Will this add two different AND conditions when,
Caller = Domain Admin
isRecursive = False
domainId = ID of a child domain

**Review by shwstppr — COMMENTED** [review]
Some suggestions @Spaceman1984 , please check

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java (hunk 3 above)
Comment (Spaceman1984) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java]: Addressed

File: api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java (hunk 4 above)
Comment (Spaceman1984) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/usage/ListUsageRecordsCmd.java]: Done

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java (hunk 5 above)
Comment (Spaceman1984) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java]: Addressed

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 6]
```diff
@@ -210,12 +212,45 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
+            } else {
+                // Domain admins can only access their own domain's usage records.
+                // Set the domain if not specified.
+                domainId = caller.getDomainId();
+            }
+
+            // Check if a domain admin is allowed to access the requested account info.
+            Account account = _accountService.getAccount(accountId);
+            Domain domain = _domainDao.findById(caller.getDomainId());
+            _accountService.checkAccess(account, domain);
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:234]: Not sure if this check is correct. Probably what you want to check is if the account belongs to the same domain or subdomain as the caller but this is checking opposite.

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 7]
```diff
@@ -234,22 +269,23 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
 
         SearchCriteria<UsageVO> sc = _usageDao.createSearchCriteria();
 
-        if (accountId != -1 && accountId != Account.ACCOUNT_ID_SYSTEM && !isAdmin && !isDomainAdmin) {
+        if (accountId != -1 && accountId != Account.ACCOUNT_ID_SYSTEM && !isAdmin && !cmd.isRecursive()) {
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:272]: @Spaceman1984 is this change needed?

**Review by shwstppr — COMMENTED** [review]
@Spaceman1984 I see a few issues. Please see the comments.
I see probable failures or wrong results:
- Domain admin calling API for an account from its subdomain
- Domain admin calling API for an account in its domain passed with accountid param while using recursive=true

**Comment by shwstppr** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java (hunk 6 above)
Comment (Spaceman1984) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:234]: This is checking if the requested account exists in the same domain as the caller. I don't see anything wrong here.

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by Spaceman1984** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@Spaceman1984 can you test if it works as before if the domain admin can list usage records for an account within the (a) same domain, (b) a sub-domain, and (c) a different domain (the domain admin is not admin of that domain).

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 8]
```diff
@@ -210,12 +212,62 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
+                Domain domain = _domainDao.findById(domainId);
+                _accountService.checkAccess(callerAccount, domain);
+            } else {
+                // Domain admins can only access their own domain's usage records.
+                // Set the domain if not specified.
+                domainId = caller.getDomainId();
+            }
+
+            // Check if a domain admin is allowed to access the requested account info.
+            Account account = _accountService.getAccount(accountId);
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:232]: @Spaceman1984 should we do a checkAccess() here as well with the provided caller and account?

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 9]
```diff
@@ -234,22 +286,27 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
-
-        if (isDomainAdmin) {
-            SearchCriteria<DomainVO> sdc = _domainDao.createSearchCriteria();
-            sdc.addOr("path", SearchCriteria.Op.LIKE, _domainDao.findById(caller.getDomainId()).getPath() + "%");
-            List<DomainVO> domains = _domainDao.search(sdc, null);
-            List<Long> domainIds = new ArrayList<Long>();
-            for (DomainVO domain : domains)
-                domainIds.add(domain.getId());
-            sc.addAnd("domainId", SearchCriteria.Op.IN, domainIds.toArray());
+        if (accountId != -1 && accountId != Account.ACCOUNT_ID_SYSTEM && !isAdmin) {
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:289]: Please simplify this.

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 10]
```diff
@@ -234,22 +286,27 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
+                sc.addAnd("accountId", SearchCriteria.Op.EQ, accountId);
+            }
         }
 
         if (domainId != null) {
-            sc.addAnd("domainId", SearchCriteria.Op.EQ, domainId);
+            if (cmd.isRecursive()) {
+                SearchCriteria<DomainVO> sdc = _domainDao.createSearchCriteria();
+                sdc.addOr("path", SearchCriteria.Op.LIKE, _domainDao.findById(domainId).getPath() + "%");
+                List<DomainVO> domains = _domainDao.search(sdc, null);
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:301]: Simplify, reuse existing DomainDao::findAllChildren

**Comment by Spaceman1984** [comment]
> @Spaceman1984 can you test if it works as before if the domain admin can list usage records for an account within the (a) same domain, (b) a sub-domain, and (c) a different domain (the domain admin is not admin of that domain).

I have already tested these scenarios. The (c) case should fail.

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java (hunk 8 above)
Comment (Spaceman1984) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:232]: I don't understand, that is being checked inside that block using checkAccess().

Comment (Spaceman1984) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:232]: @rhtyd Using checkAccess() here will result in querying the database for every child domain. I'm querying the database once by asking for all the child domains of the current domain and checking if the requested account belongs to anyone of them.

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java (hunk 9 above)
Comment (Spaceman1984) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:289]: I don't understand, how do you mean?

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java (hunk 10 above)
Comment (Spaceman1984) [reply server/src/main/java/com/cloud/usage/UsageServiceImpl.java:301]: @rhtyd Using findAllChildren here would not pass the top domain.

**Review by rhtyd — APPROVED** [review]

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 11]
```diff
@@ -210,12 +212,62 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
+            if (!matchFound) {
+                    throw new PermissionDeniedException("Domain admins may only retrieve usage records for accounts in their own domain and child domains.");
+            }
+        }
+
+        // By default users do not have access to this API.
+        // Adding checks here in case someone changes the default access.
+        if (isNormalUser) {
+            // A user can only access their own account records
+            if (caller.getId() != accountId) {
```
Comment (rhtyd) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:258]: caller.getAccountId() (in case getAccountId is not same as getId)

**Comment by rhtyd** [comment]
Consider if you can reuse existing isrecursive/domainid handler code (see service layer that handles listVolumes, listVMs... etc).

**Comment by Spaceman1984** [comment]
> Consider if you can reuse existing isrecursive/domainid handler code (see service layer that handles listVolumes, listVMs... etc).

@rhtyd the _accountMgr.buildACLViewSearchBuilder(sb, domainId, isRecursive, permittedAccounts, listProjectResourcesCriteria);

method is not compatible with how the records are fetched in the listUsageRecords class. 

In getUsageRecords() in the UsageServiceImpl class file the result set is built using 
`SearchCriteria<UsageVO> sc = _usageDao.createSearchCriteria();`
, we don't have a SearchBuilder object to pass to the method.

The code is not reusable.

**Comment by rhtyd** [comment]
Why not create a `sb` @Spaceman1984  and then use it to create a `sc`?

**Comment by Spaceman1984** [comment]
> Why not create a `sb` @Spaceman1984 and then use it to create a `sc`?

That would require a rewrite of the class.

**Comment by Spaceman1984** [comment]
> Why not create a `sb` @Spaceman1984 and then use it to create a `sc`?

@rhtyd the sb object needed here needs a VO class that implements ControlledEntity. UsageVO does not implement ControlledEntity.

**Comment by andrijapanicsb** [comment]
AS discussed with Daan, this is NOT a bug, but a feature, extension of the current API to provide new functionality

Moving to 4.16.

**Comment by Spaceman1984** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 12]
```diff
@@ -201,7 +202,8 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
         }
 
         boolean isAdmin = false;
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:204]:
```suggestion
        boolean isAdmin = _accountService.isRootAdmin(caller.getId());
```

Similarly as the other vars for domain admin / user.

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 13]
```diff
@@ -210,12 +212,65 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
+                            break;
+                        }
+                    }
+                }
+                if (!matchFound) {
+                    throw new PermissionDeniedException("Domain admins may only retrieve usage records for accounts in their own domain and child domains.");
+                }
+            }
+        }
+
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:256]: @Spaceman1984 is it possible to move the domain admin / normal user validations to separate private methods ?

**Comment by Spaceman1984** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware60u2

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-60u2

**Comment by DaanHoogland** [comment]
capacity problems, dismissing all blueorangutan results :(

**Comment by rhtyd** [comment]
@DaanHoogland @andrijapanicsb sorry guys, disagree with you - the PR introduces an enhancement to allow admins to get recursive records of a domain much like other APIs such as listVolumes, listVirtualMachines. Moved it to 4.15.1 milestone.

**Comment by DaanHoogland** [comment]
@rhtyd I don't care if it goes in 4.15 or not, as it is a minor enhancement. I wouldn't compare listUsageRecords to other list APIs though. Most testing done and ok. I just want to see another test run and am busy with that.
clgtm btw

File: server/src/main/java/com/cloud/usage/UsageServiceImpl.java [hunk 14]
```diff
@@ -200,22 +201,41 @@ public boolean generateUsageRecords(GenerateUsageRecordsCmd cmd) {
...
+            if (cmd.getAccountId() != null) {
+                // Check if a domain admin is allowed to access the requested account info.
+                checkDomainAdminAccountAccess(accountId, domainId);
+            }
+        }
+
+        // By default users do not have access to this API.
+        // Adding checks here in case someone changes the default access.
+        checkUserAccess(cmd, accountId, caller, isNormalUser);
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/usage/UsageServiceImpl.java:237]: @Spaceman1984 is this method call required for domain admin ?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
clgtm and tested

**Comment by rhtyd** [comment]
Is this ready for merging @DaanHoogland @Spaceman1984 ?

**Comment by andrijapanicsb** [comment]
Not until IR is done.

**Comment by rhtyd** [comment]
@andrijapanicsb are you lgtm on this?

**Comment by andrijapanicsb** [comment]
LGTM
